### ISSUE TITLE
tox-ansible lint_all should should not longer rely on molecule

### ISSUE THREAD

**Issue by ssbarnea** [tox-dev/tox-ansible#73]
I was very close on removing `lint` from molecule 3.2 release but decided to go for the less intrusive approach of making it a generic command. Still, I am inclined to believe we will remove it from the tool in next version and only keep a neutered `lint` sequence that displays a warning, telling people how to perform linting at repository level.

Running the lint 100 times does not make much sense, especially as each execution will also try to install dependencies. It is not only slow but also gives a poor output experience and I do not see any reasons why linting should not be performed as a single action at the repository level, instead of molecule scenario.

The way I see tox-ansible, is in the perfect position to drive this migration, as it runs at repository level, and already knows how to call molecule and ansible-test.

This is why I am inclined to propose we repurpose the generated `lint_all` action to one that performs linting of the repository in a smart way, auto-installing dependencies:

* If repository has a  `.pre-commit-config.yaml` we assume linting is performed by it and we only install pre-commit inside the virtualenv and run it
* Else we install `ansible-lint[community,yamllint]>=5.0` and run `ansible-lint` at root.

This means that we should be get a decent user experience for most users.

Open questions:
* What if user already has a `lint` or `linters` environment defined in his tox.ini file. This is almost of sure a sign that they have their own custom linting setup. Should we do something about this or not?

Please add your feedback on this, we really need to know what users expect and what are their desires.

**Comment by greg-hellings** [comment]
The current since 1.1.0 runs `ansible-lint -R [-r custom_rules] [-c config_file]` and `yamllint [-c yamllint-config-file]`. It then runs `flake8 .` so the whole repo is linted in a very minimal way. It's already very close to what you're proposing for the fall-back command. I definitely wouldn't be opposed to adding support for auto-discovery of a .pre-commit-config.yaml file to precede the built-in commands in precedence. That might be worth a separate tox-pre-commit plugin, though, as there are LOTS of people out there who use pre-commit. But that's an entirely separate issue.

We might want to consider renaming the auto-generated environment to indicate where it came from. "tox -e lint" seemed too common, in my opinion, to auto-generate that. But "tox -e lint_all" seems over-zealous. "OK, but you didn't lint my shell scripts. So what now?!" Maybe something like `tox -e lint_ansible` might be a better name for it, or `tox -e ansible_lint`, even. What do you think about that?

As for auto-discovering an existing environment, I think we should just change the `lint_all|lint_ansible` environment to detect existing commands, just like I just added to deps. Then, if you have a custom command written in the environment, we're not breaking that. We can just append our commands onto the list, as well as our deps. In general it's probably a good idea for us to update the commands to an append as well for all environments. That will allow the user an improved customization experience.

tl;dr
1) I think it is a good plan to add pre-commit detection
2) We might rename lint_all to lint_ansible
3) We can just append to existing commands for all environments, including lint_ansible

**Comment by greg-hellings** [comment]
`molecule` is no longer used by the `lint_all` target, so the title of this job is misleading.

Current lint commands are hard coded as
```
ansible-lint -R
yamllint
flake8 .
```

Adding pre-commit support is still desirable.

**Comment by ssbarnea** [comment]
You can remove yamllint standalone execution because ansible-lint includes run of yamllint for quite some time, no reason to run it again after.

**Comment by greg-hellings** [comment]
Is there a way to pass an explicit yamllint config location to `ansible-lint`'s invocation? I know there are default locations, but currently we allow the user to pass in a non-standard location that gets added to the yamllint invocation with the `-c` option. Personally I use custom yamllint config but store it in a submodule, so I use that option throughout my work. Before I remove the call to `yamllint` from tox-ansible I would want to be able to pass that through to `ansible-lint`.

**Comment by ssbarnea** [comment]
Nope and there are no plans to add one because it will still not work with other tools, for example vscode plugin for yamllint will only detect standard config file locations. Custom location is a recipe for problems.

If you make a PR to add support for it, I would refuse it. I only stated why the incentive for this kind of feature is very low.

**Comment by greg-hellings** [comment]
Hmm, for just that reason I personally have `skip_list: [yaml]` in my `ansible-lint` config and added this explicit invocation. Maybe I can make the explicit `yamllint` only enabled if the user has set a custom path for the yamllint configuration file. That way if they're explicitly setting the path, then they get that invocation but, if not, they don't have to see double-execution.
